File: examples/go/create_vm_test.go
package examplesgo

import (
	"context"
	"testing"
)

type receiver struct {
	protect bool
	seen    map[string]bool
	vms     int
	calls   int
}

func (r *receiver) CreateVM(_ context.Context, id string, _ VM) error {
	r.calls++
	if !r.protect || !r.seen[id] {
		r.vms++
		r.seen[id] = true
	}
	if r.calls == 1 {
		return ErrDeadline // Effect happened; response unavailable.
	}
	return nil
}

func TestRepresentativeCounterexampleAndContract(t *testing.T) {
	weak := &receiver{seen: map[string]bool{}}
	if err := RetryIndependent(context.Background(), weak, VM{Size: "small"}); err != nil || weak.vms != 2 {
		t.Fatalf("weak repeat: err=%v, VMs=%d", err, weak.vms)
	}
	strong := &receiver{protect: true, seen: map[string]bool{}}
	if err := RetrySameOperation(context.Background(), strong, VM{Size: "small"}, "P"); err != nil || strong.vms != 1 {
		t.Fatalf("same P: err=%v, VMs=%d", err, strong.vms)
	}
	changed := &receiver{protect: true, seen: map[string]bool{}}
	if err := RetryWithNewOperation(context.Background(), changed, VM{Size: "small"}, "P", "Q"); err != nil || changed.vms != 2 {
		t.Fatalf("P then Q: err=%v, VMs=%d", err, changed.vms)
	}
}

File: examples/c/create_vm_test.c
#include <assert.h>
#include <string.h>
#include "create_vm.c"

typedef struct {
    int protect;
    int calls;
    int vms;
    const char *seen[2];
    int seen_count;
} Receiver;

static VmOutcome create_vm(void *context, const char *operation_id, VmSpec spec) {
    Receiver *receiver = context;
    int known = 0;
    (void)spec;
    receiver->calls++;
    for (int i = 0; i < receiver->seen_count; i++) {
        if (operation_id && strcmp(receiver->seen[i], operation_id) == 0) known = 1;
    }
    if (!receiver->protect || !known) {
        receiver->vms++;
        if (receiver->protect && operation_id) receiver->seen[receiver->seen_count++] = operation_id;
    }
    return receiver->calls == 1 ? VM_NO_COMPLETION_RESPONSE : VM_OK;
}

int main(void) {
    VmSpec spec = {"small"};
    Receiver weak = {0};
    assert(retry_independent((VmClient){&weak, create_vm}, spec) == VM_OK);
    assert(weak.vms == 2);

    Receiver strong = {.protect = 1};
    assert(retry_same_operation((VmClient){&strong, create_vm}, spec, "P") == VM_OK);
    assert(strong.vms == 1);

    Receiver changed = {.protect = 1};
    assert(retry_with_new_operation((VmClient){&changed, create_vm}, spec, "P", "Q") == VM_OK);
    assert(changed.vms == 2);
    return 0;
}
